<a href="https://colab.research.google.com/github/marianameyer/DataScience/blob/main/3_Minicurso_Analise_de_Dados_Python_Dados_Cartao_Credito.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Desafio: 

Você trabalha em uma grande empresa de Cartão de Crédito e o diretor da empresa percebeu que o número de clientes que cancelam seus cartões tem aumentado significativamente, causando prejuízos enormes para a empresa

O que fazer para evitar isso? Como saber as pessoas que têm maior tendência a cancelar o cartão?

# O que temos:

Temos 1 base de dados com informações dos clientes, tanto clientes atuais quanto clientes que cancelaram o cartão

Download da Base de Dados: Botão na página

Referência: https://www.kaggle.com/sakshigoyal7/credit-card-customers

# Importando as Bibliotecas e Carregando o DF

In [1]:
# Montando o Drive para usar o Dataset
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
# Importando bibliotecas
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [4]:
# Carregando o df
df = pd.read_csv('/content/drive/MyDrive/Material de Cursos/Minicurso Análise de Dados - Hashtag/ClientesBanco.csv', encoding='ISO-8859-1')

# Visualização Prévia dos Dados

In [5]:
# Visualização do df
df.head()

CLIENTNUM Categoria  Idade Sexo  Dependentes           Educação  \
0  768805383   Cliente     45    M            3       Ensino Médio   
1  818770008   Cliente     49    F            5    Ensino Superior   
2  713982108   Cliente     51    M            3    Ensino Superior   
3  769911858   Cliente     40    F            4       Ensino Médio   
4  709106358   Cliente     40    M            3  Sem ensino formal   

    Estado Civil Faixa Salarial Anual Categoria Cartão  Meses como Cliente  \
0         Casado          $60K - $80K             Blue                  39   
1       Solteiro       Less than $40K             Blue                  44   
2         Casado         $80K - $120K             Blue                  36   
3  Não informado       Less than $40K             Blue                  34   
4         Casado          $60K - $80K             Blue                  21   

   ...  Inatividade 12m  Contatos 12m   Limite  Limite Consumido  \
0  ...                1             3  12691.0               777   
1  ...                1             2   8256.0               864   
2  ...                1             0   3418.0                 0   
3  ...                4             1   3313.0              2517   
4  ...                1             0   4716.0                 0   

   Limite Disponível  Mudanças Transacoes_Q4_Q1  Valor Transacoes 12m  \
0            11914.0                     1335.0                  1144   
1             7392.0                     1541.0                  1291   
2             3418.0                     2594.0                  1887   
3              796.0                     1405.0                  1171   
4             4716.0                     2175.0                   816   

   Qtde Transacoes 12m  Mudança Qtde Transações_Q4_Q1  \
0                   42                         1625.0   
1                   33                         3714.0   
2                   20                         2333.0   
3                   20                         2333.0   
4                   28                            2.5   

   Taxa de Utilização Cartão  
0                      0.061  
1                      0.105  
2                      0.000  
3                      0.760  
4                      0.000  

[5 rows x 21 columns]

In [6]:
# Dados estatísticos do df
df.describe()

CLIENTNUM         Idade   Dependentes  Meses como Cliente  \
count  1.012700e+04  10127.000000  10127.000000        10127.000000   
mean   7.391776e+08     46.325960      2.346203           35.928409   
std    3.690378e+07      8.016814      1.298908            7.986416   
min    7.080821e+08     26.000000      0.000000           13.000000   
25%    7.130368e+08     41.000000      1.000000           31.000000   
50%    7.179264e+08     46.000000      2.000000           36.000000   
75%    7.731435e+08     52.000000      3.000000           40.000000   
max    8.283431e+08     73.000000      5.000000           56.000000   

       Produtos Contratados  Inatividade 12m  Contatos 12m        Limite  \
count          10127.000000     10127.000000  10127.000000  10127.000000   
mean               3.812580         2.341167      2.455317   8631.953698   
std                1.554408         1.010622      1.106225   9088.776650   
min                1.000000         0.000000      0.000000   1438.300000   
25%                3.000000         2.000000      2.000000   2555.000000   
50%                4.000000         2.000000      2.000000   4549.000000   
75%                5.000000         3.000000      3.000000  11067.500000   
max                6.000000         6.000000      6.000000  34516.000000   

       Limite Consumido  Limite Disponível  Mudanças Transacoes_Q4_Q1  \
count      10127.000000       10127.000000               10127.000000   
mean        1162.814061        7469.139637                 108.817273   
std          814.987335        9090.685324                 354.050357   
min            0.000000           3.000000                   0.000000   
25%          359.000000        1324.500000                   0.631000   
50%         1276.000000        3474.000000                   0.736000   
75%         1784.000000        9859.000000                   0.859000   
max         2517.000000       34516.000000                3397.000000   

       Valor Transacoes 12m  Qtde Transacoes 12m  \
count          10127.000000         10127.000000   
mean            4404.086304            64.858695   
std             3397.129254            23.472570   
min              510.000000            10.000000   
25%             2155.500000            45.000000   
50%             3899.000000            67.000000   
75%             4741.000000            81.000000   
max            18484.000000           139.000000   

       Mudança Qtde Transações_Q4_Q1  Taxa de Utilização Cartão  
count                   10127.000000               10127.000000  
mean                       66.573618                   0.274894  
std                       283.988065                   0.275691  
min                         0.000000                   0.000000  
25%                         0.582000                   0.023000  
50%                         0.702000                   0.176000  
75%                         0.818000                   0.503000  
max                      3714.000000                   0.999000

In [7]:
# Informações úteis do df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10127 entries, 0 to 10126
Data columns (total 21 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   CLIENTNUM                      10127 non-null  int64  
 1   Categoria                      10127 non-null  object 
 2   Idade                          10127 non-null  int64  
 3   Sexo                           10127 non-null  object 
 4   Dependentes                    10127 non-null  int64  
 5   Educação                       10127 non-null  object 
 6   Estado Civil                   10127 non-null  object 
 7   Faixa Salarial Anual           10127 non-null  object 
 8   Categoria Cartão               10126 non-null  object 
 9   Meses como Cliente             10127 non-null  int64  
 10  Produtos Contratados           10127 non-null  int64  
 11  Inatividade 12m                10127 non-null  int64  
 12  Contatos 12m                   10127 non-null 

# "Arrumando" nosso DF

Iremos excluir dados/colunas que não agregam informação para a resolução do problema proposto.

In [9]:
# Exclusão da coluna CLIENTNUM 
df = df.drop('CLIENTNUM', axis=1)
df.head()

Categoria  Idade Sexo  Dependentes           Educação   Estado Civil  \
0   Cliente     45    M            3       Ensino Médio         Casado   
1   Cliente     49    F            5    Ensino Superior       Solteiro   
2   Cliente     51    M            3    Ensino Superior         Casado   
3   Cliente     40    F            4       Ensino Médio  Não informado   
4   Cliente     40    M            3  Sem ensino formal         Casado   

  Faixa Salarial Anual Categoria Cartão  Meses como Cliente  \
0          $60K - $80K             Blue                  39   
1       Less than $40K             Blue                  44   
2         $80K - $120K             Blue                  36   
3       Less than $40K             Blue                  34   
4          $60K - $80K             Blue                  21   

   Produtos Contratados  Inatividade 12m  Contatos 12m   Limite  \
0                     5                1             3  12691.0   
1                     6                1             2   8256.0   
2                     4                1             0   3418.0   
3                     3                4             1   3313.0   
4                     5                1             0   4716.0   

   Limite Consumido  Limite Disponível  Mudanças Transacoes_Q4_Q1  \
0               777            11914.0                     1335.0   
1               864             7392.0                     1541.0   
2                 0             3418.0                     2594.0   
3              2517              796.0                     1405.0   
4                 0             4716.0                     2175.0   

   Valor Transacoes 12m  Qtde Transacoes 12m  Mudança Qtde Transações_Q4_Q1  \
0                  1144                   42                         1625.0   
1                  1291                   33                         3714.0   
2                  1887                   20                         2333.0   
3                  1171                   20                         2333.0   
4                   816                   28                            2.5   

   Taxa de Utilização Cartão  
0                      0.061  
1                      0.105  
2                      0.000  
3                      0.760  
4                      0.000

Não encontrei colunas com muitos valores NaN ou que não fossem pertinentes ao problema proposto. A única exclusão é um valor faltante na coluna 'Categoria Cartão'.

Como falta um dado (de uma única linha) em um df com 10127 dados, excluirei essa linha.

In [10]:
# Exclusão de linha em que falta dado
df = df.dropna()
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10126 entries, 0 to 10126
Data columns (total 20 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Categoria                      10126 non-null  object 
 1   Idade                          10126 non-null  int64  
 2   Sexo                           10126 non-null  object 
 3   Dependentes                    10126 non-null  int64  
 4   Educação                       10126 non-null  object 
 5   Estado Civil                   10126 non-null  object 
 6   Faixa Salarial Anual           10126 non-null  object 
 7   Categoria Cartão               10126 non-null  object 
 8   Meses como Cliente             10126 non-null  int64  
 9   Produtos Contratados           10126 non-null  int64  
 10  Inatividade 12m                10126 non-null  int64  
 11  Contatos 12m                   10126 non-null  int64  
 12  Limite                         10126 non-null 

In [11]:
# Nova visualização dos dados estatísticos com os dados que serão usados
df.describe()

Idade   Dependentes  Meses como Cliente  Produtos Contratados  \
count  10126.000000  10126.000000        10126.000000          10126.000000   
mean      46.327079      2.346139           35.928995              3.812463   
std        8.016420      1.298956            7.986593              1.554440   
min       26.000000      0.000000           13.000000              1.000000   
25%       41.000000      1.000000           31.000000              3.000000   
50%       46.000000      2.000000           36.000000              4.000000   
75%       52.000000      3.000000           40.000000              5.000000   
max       73.000000      5.000000           56.000000              6.000000   

       Inatividade 12m  Contatos 12m        Limite  Limite Consumido  \
count     10126.000000  10126.000000  10126.000000      10126.000000   
mean          2.341300      2.455264   8631.962088       1162.764369   
std           1.010584      1.106267   9089.225428        815.012238   
min           0.000000      0.000000   1438.300000          0.000000   
25%           2.000000      2.000000   2555.000000        358.000000   
50%           2.000000      2.000000   4549.000000       1276.000000   
75%           3.000000      3.000000  11067.750000       1784.000000   
max           6.000000      6.000000  34516.000000       2517.000000   

       Limite Disponível  Mudanças Transacoes_Q4_Q1  Valor Transacoes 12m  \
count       10126.000000               10126.000000          10126.000000   
mean         7469.197719                 108.713167           4404.391764   
std          9091.132356                 353.912795           3397.157925   
min             3.000000                   0.000000            510.000000   
25%          1324.250000                   0.631000           2156.000000   
50%          3473.000000                   0.736000           3899.500000   
75%          9860.000000                   0.859000           4741.000000   
max         34516.000000                3397.000000          18484.000000   

       Qtde Transacoes 12m  Mudança Qtde Transações_Q4_Q1  \
count         10126.000000                   10126.000000   
mean             64.861841                      66.579995   
std              23.471594                     284.001363   
min              10.000000                       0.000000   
25%              45.000000                       0.582000   
50%              67.000000                       0.702000   
75%              81.000000                       0.818000   
max             139.000000                    3714.000000   

       Taxa de Utilização Cartão  
count               10126.000000  
mean                    0.274901  
std                     0.275704  
min                     0.000000  
25%                     0.023000  
50%                     0.175500  
75%                     0.503000  
max                     0.999000

# Análises:

## Relação entre Clientes e Cancelados:

In [13]:
# Verificação de quantos valores distintos há na coluna 'Categoria'
qt_clientes = df['Categoria'].value_counts()
qt_clientes

Cliente      8499
Cancelado    1627
Name: Categoria, dtype: int64

In [14]:
# Faremos o mesmo para obter o percentual de clientes e cancelados
qt_clientes_p = df['Categoria'].value_counts(normalize=True)
qt_clientes_p

Cliente      0.839325
Cancelado    0.160675
Name: Categoria, dtype: float64